File: Swift/LeetCode/LeetCode/StringToIntegerAtoi.swift
//
//  StringToIntegerAtoi.swift
//  LeetCode
//
//  Created by 黎赵太郎 on 25/11/2017.
//  Copyright © 2017 lizhaotailang. All rights reserved.
//
//  Implement atoi to convert a string to an integer.
//
//  Hint: Carefully consider all possible input cases. If you want a challenge,
//  please do not see below and ask yourself what are the possible input cases.
//
//  Notes: It is intended for this problem to be specified vaguely (ie, no given input specs).
//  You are responsible to gather all the input requirements up front.
//
//  Accepted. See [StringToIntegerAtoiTests](./LeetCodeTests/StringToIntegerAtoiTests.swift) for the test cases.
//

import Foundation

class StringToIntegerAtoi {
    
    func myAtoi(_ str: String) -> Int {
        let s = str.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty {
            return 0
        }
        
        if s.count == 1 {
            let value = s.unicodeScalars.first!.value
            if value <= "0".unicodeScalars.first!.value || value >= "9".unicodeScalars.first!.value {
                return 0
            }
            return Int(value - "0".unicodeScalars.first!.value)
        }
        
        let intMax = 2147483647
        let intMin = -2147483648
        let plus = s.first == Character("+")
        let minus = s.first == Character("-")
        
        var result = 0
        var startIndex = 0
        if plus || minus {
            startIndex = 1
        }
        
        let chars: [Character] = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
        let inputCharArray = Array(s)
        for i in startIndex...(inputCharArray.count - 1) {
            if chars.contains(inputCharArray[i]) {
                if (intMax / 10 - (Int(inputCharArray[i].unicodeScalars.first!.value) -  Int("0".unicodeScalars.first!.value))) <= result {
                    if (minus && (result * 10) + (Int(inputCharArray[i].unicodeScalars.first!.value) -  Int("0".unicodeScalars.first!.value)) == intMax) {
                        return -intMax
                    }
                    if minus {
                        return intMin
                    }
                    return intMax
                }
                result = result * 10 + (Int(inputCharArray[i].unicodeScalars.first!.value) - Int("0".unicodeScalars.first!.value))
            } else {
                if minus {
                    return -result
                }
                return result
            }
        }
        
        if minus {
            return -result
        }
        return result
    }
    
}

File: Golang/StringToIntegerAtoi/StringToIntegerAtoi.go
package StringToIntegerAtoi

//Implement atoi to convert a string to an integer.
//
//Hint: Carefully consider all possible input cases. If you want a challenge,
//please do not see below and ask yourself what are the possible input cases.
//
//Notes: It is intended for this problem to be specified vaguely (ie, no given input specs).
//You are responsible to gather all the input requirements up front.
//
//Accepted.

import (
	"strconv"
	"strings"
)

const intMaxValue = 2147483647
const intMinValue = -2147483648

func myAtoi(str string) int {
	s := strings.Trim(str, " ")
	stringLength := len(s)
	if stringLength == 0 {
		return 0
	}
	if stringLength == 1 {
		if s[0] <= '0' || s[0] >= '9' {
			return 0
		}
		r, _ := strconv.Atoi(s)
		return r
	}

	plus, minus := s[0] == '+', s[0] == '-'

	result, startIndex := 0, 0
	if plus || minus {
		startIndex = 1
	}

	for i := startIndex; i < stringLength; i++ {
		if s[i] >= '0' && s[i] <= '9' {
			if intMaxValue/10-int(s[i]-'0') <= result {
				if minus && result*10+int(s[i]-'0') == intMaxValue {
					return -intMaxValue
				}
				if minus {
					return intMinValue
				}
				return intMaxValue
			}
			result = result*10 + int(s[i]-'0')
		} else {
			if minus {
				return -result
			}
			return result
		}
	}

	if minus {
		return -result
	}
	return result

}
